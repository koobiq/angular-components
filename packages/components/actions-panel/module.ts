import { NgModule } from '@angular/core';
import { KbqActionsPanel } from './actions-panel';
import { KbqActionsPanelContainer } from './actions-panel-container';

const COMPONENTS = [
    KbqActionsPanelContainer
];

@NgModule({
    imports: COMPONENTS,
    exports: COMPONENTS,
    providers: [KbqActionsPanel]
})
export class KbqActionsPanelModule {}
