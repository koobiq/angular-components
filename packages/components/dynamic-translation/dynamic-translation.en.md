`KbqDynamicTranslation` â€” component for embedding custom components into translatable strings.

<!-- example(dynamic-translation-overview) -->

### Adding a component

Components are embedded into strings using special syntax: `[[<SLOT_NAME>:<CONTEXT>]]`, where `SLOT_NAME` is the key of the registered component, and `CONTEXT` is the data passed to the component.

<!-- prettier-ignore -->
```html
<!-- Slot with string passing -->
<kbq-dynamic-translation text="[[myLinkSlot:Open link]]">
    <a *kbqDynamicTranslationSlot="'myLinkSlot'; let context">{{ context }}</a>
</kbq-dynamic-translation>

<!-- Slot with list passing -->
<kbq-dynamic-translation text="[[myListSlot:(first,second,third)]]">
    <ul *kbqDynamicTranslationSlot="'myListSlot'; let items">
        @for (item of items; track $index) {
            <li>{{ item }}</li>
        }
    </ul>
</kbq-dynamic-translation>
```

### Key features

- Embedding components ([links](/en/components/link), [buttons](/en/components/button), [dropdowns](/en/components/dropdown), etc.) into localized strings.
- Separation of interface logic from localization texts.
- Support for multiple slots in a single string.

### When to use

- When you need to dynamically change the composition and context of embedded components depending on localization.
- If you need to add interactive elements ([links](/en/components/link), [buttons](/en/components/button), [dropdowns](/en/components/dropdown), etc.) directly into translatable text.
- For highlighting parts of text with special styles.
