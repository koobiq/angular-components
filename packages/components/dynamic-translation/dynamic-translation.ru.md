`KbqDynamicTranslation` — компонент для встраивания пользовательских компонентов в переводимые строки.

<!-- example(dynamic-translation-overview) -->

### Добавление компонента

Компонент встраивается в строку с помощью специального синтаксиса: `[[<slotName>:<context>]]`, где `slotName` — ключ зарегистрированного компонента, а `context` — данные, передаваемые в компонент.

<!-- prettier-ignore -->
```html
<!-- Слот с передачей строки -->
<kbq-dynamic-translation text="[[myLinkSlot:Открыть ссылку]]">
    <a *kbqDynamicTranslationSlot="'myLinkSlot'; let context">{{ context }}</a>
</kbq-dynamic-translation>

<!-- Слот с передачей списка -->
<kbq-dynamic-translation text="[[myListSlot:(первый,второй,третий)]]">
    <ul *kbqDynamicTranslationSlot="'myListSlot'; let context">
        @for (item of context; track $index) {
            <li>{{ item }}</li>
        }
    </ul>
</kbq-dynamic-translation>
```

### Основные возможности

- Встраивание компонентов ([ссылки](/ru/components/link), [кнопки](/ru/components/button), [выпадающие списки](/ru/components/dropdown) и др.) в локализованные строки.
- Разделение логики интерфейса от текстов локализации.
- Поддержка нескольких слотов в одной строке.

### Когда использовать

- При необходимости динамически изменять состав и контекст встраиваемых компонентов в зависимости от локализации.
- Если требуется добавить интерактивные элементы ([ссылки](/ru/components/link), [кнопки](/ru/components/button), [выпадающие списки](/ru/components/dropdown) и др.) прямо в переводимый текст.
- Для выделения части текста специальными стилями.
