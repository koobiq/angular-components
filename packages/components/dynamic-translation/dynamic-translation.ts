import { NgTemplateOutlet } from '@angular/common';
import {
    afterNextRender,
    ChangeDetectionStrategy,
    Component,
    computed,
    contentChildren,
    Directive,
    effect,
    ElementRef,
    inject,
    input,
    Renderer2,
    Signal,
    TemplateRef,
    viewChildren
} from '@angular/core';

/**
 * @docs-private
 */
type KbqDynamicTranslationParsedSlot = {
    type: 'text' | 'slot';
    template: TemplateRef<unknown> | null;
    context?: { $implicit: unknown };
    text?: string;
};

/**
 * @docs-private
 */
export interface KbqDynamicTranslationHelperSlot {
    name: string;
    tag: string;
    attributes?: { key: string; value: string }[];
    content?: string;
}

@Directive({
    standalone: true,
    selector: '[kbqDynamicTranslationSlot]'
})
export class KbqDynamicTranslationSlot {
    readonly slotName = input.required<string>({ alias: 'kbqDynamicTranslationSlot' });

    /**
     * @docs-private
     */
    readonly templateRef = inject(TemplateRef);
}

@Component({
    standalone: true,
    imports: [NgTemplateOutlet],
    selector: 'kbq-dynamic-translation',
    template: `
        @for (slot of parsedSlots(); track $index) {
            @switch (slot.type) {
                @case ('slot') {
                    <ng-container *ngTemplateOutlet="slot.template; context: slot.context" />
                }
                @default {
                    <ng-container>{{ slot.text }}</ng-container>
                }
            }
        }
    `,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class KbqDynamicTranslation {
    private readonly slots = contentChildren(KbqDynamicTranslationSlot);

    readonly text = input.required<string>();

    /**
     * @docs-private
     */
    protected readonly parsedSlots: Signal<KbqDynamicTranslationParsedSlot[]> = computed(() =>
        this.parseSlots(this.text(), this.slots())
    );

    private parseSlots(
        text: string,
        slots: ReadonlyArray<KbqDynamicTranslationSlot>
    ): KbqDynamicTranslationParsedSlot[] {
        const slotTemplatesByName = slots.reduce<Record<string, TemplateRef<unknown>>>(
            (map, { slotName, templateRef }) => {
                const _slotName = slotName();

                if (_slotName) map[_slotName] = templateRef;

                return map;
            },
            {}
        );
        const parsedSlots: KbqDynamicTranslationParsedSlot[] = [];
        const slotSelector = new RegExp(`\\[\\[(${Object.keys(slotTemplatesByName).join('|')})(?::(.+?))?\\]\\]`, 'g');
        let match: RegExpExecArray | null;
        let lastIndex = 0;

        while ((match = slotSelector.exec(text)) !== null) {
            const [slotMatch, slotName, slotText] = match;
            const [listMatch, listString] = new RegExp('^\\((.*?)\\)$', 'g').exec(slotText) ?? [];
            const textBeforeSlot = text.substring(lastIndex, +match.index);

            if (textBeforeSlot) {
                parsedSlots.push({ type: 'text', text: textBeforeSlot, template: null });
            }

            parsedSlots.push({
                type: 'slot',
                template: slotTemplatesByName[slotName] || null,
                context: { $implicit: listMatch && listString ? listString.split(',') : slotText }
            });

            lastIndex = match.index + slotMatch.length;
        }

        const textAfterLastSlot = text.substring(lastIndex);

        if (textAfterLastSlot) {
            parsedSlots.push({ type: 'text', text: textAfterLastSlot, template: null });
        }

        return parsedSlots;
    }
}

/**
 * @docs-private
 */
@Component({
    standalone: true,
    imports: [KbqDynamicTranslation, KbqDynamicTranslationSlot],
    selector: 'kbq-dynamic-translation-helper',
    template: `
        <kbq-dynamic-translation [text]="text()">
            @for (slot of slots(); track $index) {
                <ng-container *kbqDynamicTranslationSlot="slot.name; let content">
                    <span
                        #replaceableSlotContainer
                        [attr.data-slot-name]="slot.name"
                        [attr.data-slot-content]="slot.content || content"
                    ></span>
                </ng-container>
            }
        </kbq-dynamic-translation>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class KbqDynamicTranslationHelper {
    private readonly renderer = inject(Renderer2);
    private readonly replaceableSlotContainers = viewChildren<ElementRef<HTMLElement>>('replaceableSlotContainer');
    private readonly slotsByName = new Map<string, KbqDynamicTranslationHelperSlot>();

    readonly text = input.required<string>();

    readonly slots = input.required<KbqDynamicTranslationHelperSlot[]>();

    constructor() {
        effect(() => {
            this.slotsByName.clear();
            this.slots().forEach((slot) => this.slotsByName.set(slot.name, slot));
        });

        afterNextRender(() => {
            this.replaceSlots();
        });
    }

    private replaceSlots(): void {
        this.replaceableSlotContainers().forEach(({ nativeElement }: ElementRef<HTMLElement>) => {
            const slotName = nativeElement.getAttribute('data-slot-name')!;
            const slotContent = nativeElement.getAttribute('data-slot-content')!;
            const { tag, attributes } = this.slotsByName.get(slotName)!;
            const elementToRender = this.renderer.createElement(tag);

            this.renderer.appendChild(elementToRender, this.renderer.createText(slotContent));

            attributes?.forEach(({ key, value }) => this.renderer.setAttribute(elementToRender, key, value));

            nativeElement.replaceWith(elementToRender);
        });
    }
}
