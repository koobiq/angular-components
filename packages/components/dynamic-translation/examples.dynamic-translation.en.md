### Dynamic component creation

<!-- example(dynamic-translation-with-dynamic-component-creation) -->
