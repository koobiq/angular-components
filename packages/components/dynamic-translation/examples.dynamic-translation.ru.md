### Динамическое создание компонента

<!-- example(dynamic-translation-with-dynamic-component-creation) -->
