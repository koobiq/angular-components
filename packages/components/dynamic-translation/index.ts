export * from './dynamic-translation';
export * from './module';
