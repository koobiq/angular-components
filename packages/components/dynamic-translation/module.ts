import { NgModule } from '@angular/core';
import { KbqDynamicTranslation, KbqDynamicTranslationHelper, KbqDynamicTranslationSlot } from './dynamic-translation';

const COMPONENTS = [
    KbqDynamicTranslationSlot,
    KbqDynamicTranslationHelper,
    KbqDynamicTranslation
];

@NgModule({
    imports: COMPONENTS,
    exports: COMPONENTS
})
export class KbqDynamicTranslationModule {}
