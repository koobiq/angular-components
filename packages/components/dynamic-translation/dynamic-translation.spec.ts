import { ChangeDetectionStrategy, Component, DebugElement, model, Provider, Type } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { KbqLinkModule } from '../link';
import { KbqDynamicTranslation } from './dynamic-translation';
import { KbqDynamicTranslationModule } from './module';

const createComponent = <T>(component: Type<T>, providers: Provider[] = []): ComponentFixture<T> => {
    TestBed.configureTestingModule({
        imports: [component, NoopAnimationsModule],
        providers
    });
    const fixture = TestBed.createComponent<T>(component);

    fixture.autoDetectChanges();

    return fixture;
};

const getDynamicTranslationDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqDynamicTranslation));
};

@Component({
    standalone: true,
    imports: [KbqDynamicTranslationModule, KbqLinkModule],
    selector: 'test-dynamic-translation',
    template: `
        <kbq-dynamic-translation [text]="text()">
            <a *kbqDynamicTranslationSlot="'testSlot1'; let context" kbq-link href="#">{{ context }}</a>
            <strong *kbqDynamicTranslationSlot="'testSlot2'; let context">{{ context }}</strong>
            <ul *kbqDynamicTranslationSlot="'testListSlot'; let list">
                @for (item of list; track $index) {
                    <li>{{ item }}</li>
                }
            </ul>
        </kbq-dynamic-translation>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class TestDynamicTranslation {
    readonly text = model(
        'text before slot1 [[testSlot1:slot1 context]] text between slots [[testSlot2:slot2 context]] text after slot2'
    );
}

describe(KbqDynamicTranslationModule.name, () => {
    it('should render component with 2 slots', () => {
        const { debugElement } = createComponent(TestDynamicTranslation);

        expect(getDynamicTranslationDebugElement(debugElement).nativeNode).toMatchSnapshot();
    });

    it('should change text dynamically', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;

        componentInstance.text.set('NEW text before slot [[testSlot1:NEW slot context]] NEW text after slot');
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeNode).toMatchSnapshot();
    });

    it('should render text without slots', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;
        const text = 'text without any slot';

        componentInstance.text.set(text);
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeElement.textContent).toBe(text);
    });

    it('should handle empty text input', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;

        componentInstance.text.set('');
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeElement.textContent).toBe('');
    });

    it('should handle slots without context', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;

        componentInstance.text.set('with empty slots [[testSlot1]]');
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeNode).toMatchSnapshot();
    });

    it('should handle multiple same slots', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;

        componentInstance.text.set(
            'first slot [[testSlot2:first context]] and second slot [[testSlot2:second context]]'
        );
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeNode).toMatchSnapshot();
    });

    it('should handle list-like slot context (parentheses)', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;

        componentInstance.text.set('with list [[testListSlot:(item1,item2,item3)]]');
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeNode).toMatchSnapshot();
    });

    it('should handle special characters', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;

        componentInstance.text.set(
            'with special chars [[testSlot2:0123456789!@#$%^&*()_+-=[]{}]] and 0123456789!@#$%^&*()_+-=[]{}'
        );
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeNode).toMatchSnapshot();
    });

    it('should handle malformed slot syntax', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;

        componentInstance.text.set('with [[incomplete and [invalid] and [[testSlot2:valid]] syntax');
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeNode).toMatchSnapshot();
    });

    it('should handle unicode characters', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;

        componentInstance.text.set('with unicode [[testSlot2:тест 测试 🎉]]');
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeNode).toMatchSnapshot();
    });

    it('should handle undefined slot', () => {
        const fixture = createComponent(TestDynamicTranslation);
        const { debugElement, componentInstance } = fixture;

        componentInstance.text.set('with undefined slot [[undefinedSlot:undefined slot context]]');
        fixture.detectChanges();

        expect(getDynamicTranslationDebugElement(debugElement).nativeNode).toMatchSnapshot();
    });
});
