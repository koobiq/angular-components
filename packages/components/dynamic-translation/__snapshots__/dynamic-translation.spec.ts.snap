// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KbqDynamicTranslationModule should change text dynamically 1`] = `
<kbq-dynamic-translation>
  
  NEW text before slot 
  
  
  <a
    class="kbq-link kbq-text-only"
    href="#"
    kbq-link=""
    tabindex="0"
  >
    NEW slot context
  </a>
  
  
  
  
   NEW text after slot
  
  
  
</kbq-dynamic-translation>
`;

exports[`KbqDynamicTranslationModule should handle list-like slot context (parentheses) 1`] = `
<kbq-dynamic-translation>
  
  with list 
  
  
  <ul>
    <li>
      item1
    </li>
    <li>
      item2
    </li>
    <li>
      item3
    </li>
    
  </ul>
  
  
  
  
</kbq-dynamic-translation>
`;

exports[`KbqDynamicTranslationModule should handle malformed slot syntax 1`] = `
<kbq-dynamic-translation>
  
  with [[incomplete and [invalid] and 
  
  
  <strong>
    valid
  </strong>
  
  
  
  
   syntax
  
  
  
</kbq-dynamic-translation>
`;

exports[`KbqDynamicTranslationModule should handle multiple same slots 1`] = `
<kbq-dynamic-translation>
  
  first slot 
  
  
  <strong>
    first context
  </strong>
  
  
  
  
   and second slot 
  
  
  <strong>
    second context
  </strong>
  
  
  
  
</kbq-dynamic-translation>
`;

exports[`KbqDynamicTranslationModule should handle slots without context 1`] = `
<kbq-dynamic-translation>
  
  with empty slots 
  
  
  <a
    class="kbq-link kbq-text-only"
    href="#"
    kbq-link=""
    tabindex="0"
  >
    
  </a>
  
  
  
  
</kbq-dynamic-translation>
`;

exports[`KbqDynamicTranslationModule should handle special characters 1`] = `
<kbq-dynamic-translation>
  
  with special chars 
  
  
  <strong>
    0123456789!@#$%^&*()_+-=[]{}
  </strong>
  
  
  
  
   and 0123456789!@#$%^&*()_+-=[]{}
  
  
  
</kbq-dynamic-translation>
`;

exports[`KbqDynamicTranslationModule should handle unicode characters 1`] = `
<kbq-dynamic-translation>
  
  with unicode 
  
  
  <strong>
    —Ç–µ—Å—Ç ÊµãËØï üéâ
  </strong>
  
  
  
  
</kbq-dynamic-translation>
`;

exports[`KbqDynamicTranslationModule should render component with 2 slots 1`] = `
<kbq-dynamic-translation>
  
  text before slot1 
  
  
  <a
    class="kbq-link kbq-text-only"
    href="#"
    kbq-link=""
    tabindex="0"
  >
    slot1 context
  </a>
  
  
  
  
   text between slots 
  
  
  <strong>
    slot2 context
  </strong>
  
  
  
  
   text after slot2
  
  
  
</kbq-dynamic-translation>
`;
